jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+                    "Transaction: " + txid + " has a conflict, abandoning.")
+                continue
+
+            removed_utxos, added_utxos = self.wallet.process_new_tx(txd, txid)
+
+            # first fire 'all' type callbacks, irrespective of if the
+            # transaction pertains to anything known (but must
+            # have correct label per above):
+            for f in self.callbacks["all"]:
+                # note we need no return value as we will never
+                # remove these from the list
+                f(txd, txid)
+
+            # The tuple given as the second possible key for the dict
+            # is such because dict keys must be hashable types, so a simple
+            # replication of the entries in the list tx["outs"], where tx
+            # was generated via jmbitcoin.deserialize, is unacceptable to
+            # Python, since they are dicts. However their keyset is deterministic
+            # so it is sufficient to convert these dicts to tuples with fixed
+            # ordering, thus it can be used as a key into the self.callbacks
+            # dicts. (This is needed because txid is not always available
+            # at the time of callback registration).
+            possible_keys = [txid, tuple(
+                    (x["script"], x["value"]) for x in txd["outs"])]
+
+            # note that len(added_utxos) > 0 is not a sufficient condition for
+            # the tx being new, since wallet.add_new_utxos will happily re-add
+            # a utxo that already exists; but this does not cause re-firing
+            # of callbacks since we in these cases delete the callback after being
+            # called once.
+            # Note also that it's entirely possible that there are only removals,
+            # not additions, to the utxo set e.g. in tumbler sweeps or direct
+            # send payments.
+            if len(added_utxos) > 0 or len(removed_utxos) > 0:
+                if confs == 0:
+                    for k in possible_keys:
+                        if k in self.callbacks["unconfirmed"]:
+                            for f in self.callbacks["unconfirmed"][k]:
+                                # True implies success, implies removal:
+                                if f(txd, txid):
+                                    self.callbacks["unconfirmed"][k].remove(f)
+                                    # keep monitoring for conf > 0:
+                                    self.active_txids.append(txid)
+                elif confs > 0:
+                    for k in possible_keys:
+                        if k in self.callbacks["confirmed"]:
+                            for f in self.callbacks["confirmed"][k]:
+                                if f(txd, txid, confs):
+                                    self.callbacks["confirmed"][k].remove(f)
+                                    if txid in self.active_txids:
+                                        self.active_txids.remove(txid)
```

Comment:
Entries from `active_txids` should also be removed if there was no corresponding callback.

Reply:
Don't think so.

The dict `confirm_callbacks` is keyed by txinfo, either the txid expected (for the Taker, who knows it) or a tuple of outputs (for Makers, who don't , when they set up the callback), and we use active_txids as a way of keeping this transaction as one we're listening on, even after we've processed it once (at confs = 0).

If we call f(txd, txid, confs) and it returns False it means that that transaction was not applicable to that callback f, so nothing is to be done because we're still waiting for the right f to be called.

Reply:
I don't see any part of code where it's ensured that a callback must exist for anything added to `active_txids`. So this might lead to a memleak.

Reply:
I see (to be clear, although I'm sure you get it: active_txids is only added to on a successful return from an unconfirm callback).

So as it stands it relies on the fact that the caller/user of WalletService adds a confirm callback matching each unconfirm callback, else the list can have dangling items where an active_txid entry is created from unconf, but no conf callback ever happens. Obviously it *is* the case, as it stands, that we always add both in JM, but they can time out and never get called. Btw a very similar thing existed in the old version of the code (there were dangling callbacks that just hung around).

Considering fixing it, the trickiness would be if you wanted to check whether any callback referenced a txid, you'd need to make sure each callback was indexed/referenced with a txid, but when they're set up, some of them aren't able to do that, but can only index by tx-out-set. They could be altered at the time of unconf callback to do that. But it is going to be a chunk of extra logic, I'm not sure it's worth doing that? (the logic of the transaction monitor is already quite nested and conditional).

To me a dangling txid in that list for every tx that doesn't get confirmed in 6 hours is basically nothing. But I agree it could be made more watertight.

Reply:
It's not super important. Maybe put a FIXME comment in the code somewhere and leave it alone until someone has trouble with it.

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/248dbbad63bd27862e7b3b9354468db13d09aec3